duktape/src/main/jni/JavaMethod.cpp

```diff
@@ -0,0 +1,132 @@
+/*
+ * Copyright (C) 2016 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include "JavaMethod.h"
+#include <algorithm>
+#include "JString.h"
+
+namespace {
+
+JavaMethod::Type getType(JNIEnv* env, jobject typeObject) {
+  if (env->IsSameObject(typeObject, env->FindClass("java/lang/String"))) {
+    return JavaMethod::String;
+  }
+  // TODO: support some more types.
+  return JavaMethod::Void;
+}
+
+/**
+ * Convert the duktape value at the top of the stack to the given Java type.
+ * throws a JavaScript TypeError if the JS type is not {@code type}.
+ */
+jvalue popJavaType(JNIEnv* env, duk_context* ctx, JavaMethod::Type type) {
+  const duk_idx_t index = -1;
+  jvalue value;
+  switch (type) {
+    default:
+    case JavaMethod::Void:
+      duk_error(ctx, DUK_ERR_API_ERROR, "unsupported argument type: %d", type);
+
+    case JavaMethod::String:
+      // Check if the caller passed in a null string.
+      value.l = duk_get_type(ctx, index) != DUK_TYPE_NULL
+          ? env->NewStringUTF(duk_require_string(ctx, index))
+          : nullptr;
+      break;
+  }
+  duk_pop(ctx);
+  return value;
+}
+
+/**
+ * Convert the given {@code value} to a duktape type and push it onto the stack (for a function
+ * call return).
+ */
+duk_ret_t pushJavaScriptType(JNIEnv* env, duk_context *ctx, JavaMethod::Type type, jvalue value) {
+  switch (type) {
+    case JavaMethod::Void:
+      return 0;
+
+    case JavaMethod::String: {
+      const JString result(env, static_cast<jstring>(value.l));
+      duk_push_string(ctx, result);
+      return 1;
```

Comment:
Yes, the return value to duktape is the number of entries pushed to the stack.  I'll add a comment.